proposed/resource-limit.md

```diff
@@ -0,0 +1,868 @@
+
+For example, this will likely involve a modified `ResourceLease` (expand details for API)
+
+<details>
+
+```c#
+public struct ResourceLease : IDisposable
+{
+    // This represents whether resource acquisition was successful
+    public bool IsAcquired { get; }
+
+    // This represents the count of resources obtained in the lease
+    public int Count { get; }
+
+    // This represents additional metadata that can be returned as part of a call to Acquire/AcquireAsync
+    // Potential uses could include a RetryAfter value or an error code.
+    public object? State { get; }
+
+    // Private fields to be used by `onDispose()`, this is not a public API shown here for completeness
+    private ResourceLimiter? _resourceLimiter;
+    private Action<ResourceLimiter?, int, object?>? _onDispose;
+
+    // Constructor which sets all the readonly values
+    public ResourceLease(
+        bool isAcquired, 
+        int Count, 
+        object? state, 
+        ResourceLimiter? resourceLimiter, 
+        Action<ResourceLimiter?, int, object?>? onDispose);
+
+    // Release a specific amount of resources, if possible
+
+    public void Release(int releaseCount);
+
+    // Return the acquired resources, calls onDispose with ResourceLimiter
+    // This can only be called once, it's an user error if called more than once
+    public void Dispose();
+}
+```
+
+</details>
+
+
+#### Aggregated limiters
+
+A separate abstraction different from the one proposed here is required to support scenarios where aggregated resources are needed.
+
+For high cardinality resources such as rate limit by IP, we don't want to have a rate limit per bucket (i.e. one rate limiter per IP Address). As such, we need an API where you can pass in a resourceID. This is in contrast to simpler resources where a key is not necessary, such as a n requests/second limit, where requiring a default key to be passed in becomes awkward. Hence the simpler API proposed here is preferred.
+
+However, we have not yet found any use cases for aggregated limiters in dotnet/runtime and hence it's not part of this proposal. If we eventually deem that such aggregated limiters are needed in the BCL as well, it can be added independently from the APIs proposed here.
+
+Sample API:
+
+```c#
+// Represent an aggregated resource (e.g. a resource limiter aggregated by IP)
+public abstract class AggregatedResourceLimiter<TKey>
+{
+    // an inaccurate view of resources
+    public abstract int EstimatedCount(TKey resourceID);
+
+    // Fast synchronous attempt to acquire resources
+    public abstract ResourceLease Acquire(TKey resourceID, int requestedCount);
```

Comment:
I think the ResourceLease needs to be generic so it can store the `AggregatedResourceLimiter<TKey>`